src/parser.js

```diff
@@ -6,6 +6,55 @@ import decodeB64ToUint8Array from '@videojs/vhs-utils/es/decode-b64-to-uint8-arr
 import LineStream from './line-stream';
 import ParseStream from './parse-stream';
 
+// set SERVER-CONTROL hold back based upon targetDuration and partTargetDuration
+// we need this helper because defaults are based upon targetDuration and
+// partTargetDuration being set, but they may not be if SERVER-CONTROL appears before
+// target durations are set.
+const setHoldBack = function(manifest) {
+  const {serverControl, targetDuration, partTargetDuration} = manifest;
+
+  if (!serverControl) {
+    return;
+  }
+
+  const tag = '#EXT-X-SERVER-CONTROL';
+  const hb = 'HOLD-BACK';
+  const phb = 'PART-HOLD-BACK';
+  const minTargetDuration = targetDuration && targetDuration * 3;
+  const minPartDuration = partTargetDuration && partTargetDuration * 2;
+
+  if (targetDuration && !serverControl.hasOwnProperty(hb)) {
+    serverControl[hb] = minTargetDuration;
+    this.trigger('info', {
+      message: `${tag} defaulting ${hb} to targetDuration * 3 (${minTargetDuration}).`
+    });
+  }
+
+  if (minTargetDuration && serverControl[hb] < minTargetDuration) {
+    this.trigger('warn', {
+      message: `${tag} clamping ${hb} (${serverControl[hb]}) to targetDuration * 3 (${minTargetDuration})`
+    });
+    serverControl[hb] = minTargetDuration;
+  }
+
+  // default no part hold back to part target duration * 3
+  if (partTargetDuration && !serverControl.hasOwnProperty(phb)) {
+    serverControl[phb] = partTargetDuration * 3;
+    this.trigger('info', {
```

Comment:
should we have this as `warn` if we have an `EXT-X-PART-INF` tag in the manifest?

Reply:
It's valid for it not to exist, the spec just specifies a default to use.

Reply:
The spec says it's required if if PART-INF is present (end of part-hold-back section in https://tools.ietf.org/html/draft-pantos-hls-rfc8216bis-08#section-4.4.3.8)
>       PART-HOLD-BACK is REQUIRED if the Playlist contains the EXT-X-PART-INF tag.

Reply:
ahh I see the confusion here now. partTargetDuration maps to `PART-TARGET` the only attribute of `EXT-X-PART-INF` we can warn if we have a `manifest.partInf` without a `PART-TARGET`, but we already do that for `EXT-X-PART-INF` and it seems like we would be logging the same thing twice.